### PR TITLE
Add a test suite

### PR DESCRIPTION
**Pull request by dscho**
The GitForWindowsHelper GitHub App is becoming too vital for the Git for Windows project to risk deploying code without a test suite. So let's add a test suite that is run as part of each Pull Request.

Unsurprisingly, in the process of writing this test suite a couple of bugs were discovered and subsequently fixed, proving the value of having this test suite right off the bat.

This closes https://github.com/git-for-windows/gfw-helper-github-app/issues/16

### CODE REVIEW

**Comment by rimrul** [comment]
Does the `__tests_` directory intentionally have two leading underscores but only one trailing underscore?

**Comment by dscho** [comment]
> Does the `__tests_` directory intentionally have two leading underscores but only one trailing underscore?

No. Good point, will fix. Thank you!

**Comment by dscho** [comment]
@rimrul could I bother you with a review? Introducing the test suite unearthed a couple of bugs that I fixed in this PR, for example the one that [made the GCM `/open pr` fail](https://github.com/git-for-windows/git/issues/4309)...

File: __tests__/index.test.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        req: {
+            method: 'GET'
+        }
+    }
+
+    const expectInvalidWebhook = async (message) => {
+        context.log.mockClear()
+        expect(await index(context, context.req)).toBeUndefined()
+        expect(context.log).toHaveBeenCalledTimes(1)
+        expect(context.log.mock.calls[0][0].message).toEqual(message)
```
Comment (rimrul) [inline __tests__/index.test.js:18]: `context.log()` gets called with a single string as an argument, doesn't it? Shouldn't this be `expect(context.log.mock.calls[0][0]).toEqual(message)` then?

File: __tests__/index.test.js [hunk 2]
```diff
@@ -40,4 +41,82 @@ test('reject requests other than webhook payloads', async () => {
...
+test('test /hi', async () => {
+    const context = mockIssueComment('/hi')
+    expect(await index(context, context.req)).toBeUndefined()
+    expect(context.res).toEqual({
+        body: 'I said hi! new-comment-url-Hi @statler and waldorf!',
+        headers: undefined,
+        status: undefined
+    })
+    expect(mockGetInstallationAccessToken).toHaveBeenCalledTimes(1)
+    expect(mockGitHubApiRequestAsApp).not.toHaveBeenCalled()
```
Comment (rimrul) [inline __tests__/index.test.js:196]: Should probably also `expect(mockGithubApiRequest).toHaveBeenCalledTimes(1)`

**Review by rimrul — COMMENTED** [review]
Haven't reviewed all commots yet.

File: __tests__/index.test.js (hunk 2 above)
Comment (dscho) [reply __tests__/index.test.js:196]: Actually, there is that `.not.` part...

Comment (rimrul) [reply __tests__/index.test.js:196]: That's for `mockGithubApiRequestAsApp`, I'm talking about an additional `expect()` for `mockGithubApiRequest`.

Comment (dscho) [reply __tests__/index.test.js:196]: Oh, but I think that should not be called because that call is "mocked away". Will check in a minute.

Comment (dscho) [reply __tests__/index.test.js:197]: Ah. It _is_ called. To add the issue comment.

File: __tests__/index.test.js (hunk 1 above)
Comment (dscho) [reply __tests__/index.test.js:19]: It is a single argument, but that argument is not a string, it is an `Error`. I've added a comment to explain that.

File: __tests__/index.test.js [hunk 3]
```diff
@@ -82,8 +82,8 @@ const makeContext = (body, headers) => {
     }
 }
 
-const mockIssueComment = (comment) => {
-    return makeContext({
+const testIssueComment = (comment, fn) => {
+    const context = makeContext(extend({
```
Comment (rimrul) [inline __tests__/index.test.js:154]: `extend()` snuck in here, but only gets defined in the next commit.

Comment (dscho) [reply __tests__/index.test.js:154]: Right you are! I had not made sure that `git rebase -x "npm run test"` succeeds, sorry. Now force-pushed. The range-diff is:



* 1:  18607d5 !  1:  9e00a3e Support `npm run lint`

   ``````diff
   @@ package.json
         "jest": "^29.4.2"
    +  },
    +  "scripts": {
   -+    "lint": "eslint -c .eslintrc.json '**/*.js'",
   ++    "lint": "eslint -c .eslintrc.json '**/*.js'"
       }
     }
   ``````

* 2:  5afa4f8 =  2:  41e25fb npm run lint: handle Jest correctly
* 3:  e64cdf3 !  3:  ff496e9 jest --init

   ``````diff
   @@ jest.config.js (new)
    
     ## package.json ##
    @@
   +     "jest": "^29.4.2"
       },
       "scripts": {
   -     "lint": "eslint -c .eslintrc.json '**/*.js'",
   +-    "lint": "eslint -c .eslintrc.json '**/*.js'"
   ++    "lint": "eslint -c .eslintrc.json '**/*.js'",
    +    "test": "jest"
       }
     }
   ``````

* 4:  f5aa71e =  4:  5227939 jest.config: adjust indentation
* 5:  3d826ad =  5:  7736b7d Start the test suite
* 6:  1fa9080 =  6:  4d81bf4 Run the tests as part of the PR checks
* 7:  30373d4 =  7:  56ee27e tests: verify the remaining wehbook validations
* 8:  1dc54cf =  8:  0e36b66 tests: add a regression test for the `/hi` command
* 9:  b7b802b !  9:  b557b82 tests: show `context.log()` output upon failure

   ``````diff
   @@ __tests__/index.test.js: const makeContext = (body, headers) => {
    -const mockIssueComment = (comment) => {
    -    return makeContext({
    +const testIssueComment = (comment, fn) => {
   -+    const context = makeContext(extend({
   ++    const context = makeContext({
             action: 'created',
             comment: {
                 body: comment,
   @@ __tests__/index.test.js: const mockIssueComment = (comment) => {
         }, {
             'x-github-event': 'issue_comment'
         })
   ++
    +    test(`test ${comment}`, async () => {
    +        try {
    +            await fn(context)
   ``````

* 10:  a4b9663 ! 10:  d155922 testsIssueComment: prepare for additional body contents

   ``````diff
   @@ __tests__/index.test.js: const makeContext = (body, headers) => {
     }
     
    -const testIssueComment = (comment, fn) => {
   +-    const context = makeContext({
    +function extend (a, ...list) {
    +    for (const b of list) {
    +        for (const key of Object.keys(b)) {
   @@ __tests__/index.test.js: const makeContext = (body, headers) => {
    +    const repo = bodyExtra_?.repository?.name || 'git'
    +    const number = bodyExtra_?.issue?.number || 0
    +    const pullOrIssues = bodyExtra_?.issue?.pull_request ? 'pull' : 'issues'
   -     const context = makeContext(extend({
   ++    const context = makeContext(extend({
             action: 'created',
             comment: {
                 body: comment,
   @@ __tests__/index.test.js: const testIssueComment = (comment, fn) => {
    +    }, bodyExtra_ ? bodyExtra_ : {}), {
             'x-github-event': 'issue_comment'
         })
   -     test(`test ${comment}`, async () => {
   + 
   ``````

* 11:  e87b30a = 11:  5ae167a tests: also test `/open pr`
* 12:  af8529d = 12:  e9cbbec tests: also test `/deploy`
* 13:  c7bd4a1 = 13:  51550f3 Add a unit test for `component-updates`
* 14:  a140d03 = 14:  700b95b guess release notes: support GNU TLS tickets better
* 15:  f9c0bb5 = 15:  2ead6fe guess release notes: handle newer GCM tickets better
* 16:  afb96cc = 16:  2180943 tests: also test `/add release note`

Comment (dscho) [reply __tests__/index.test.js:154]: @rimrul does it look okay now?

File: __tests__/index.test.js [hunk 4]
```diff
@@ -154,3 +201,40 @@ testIssueComment('/hi', async (context) => {
...
+        body: `I edited the comment: appended-comment-body-existing comment body
+
+The MINGW workflow run [was started](dispatched-workflow-open-pr.yml)`,
+        headers: undefined,
+        status: undefined
+    })
+    expect(mockGetInstallationAccessToken).toHaveBeenCalledTimes(1)
+    expect(mockGitHubApiRequestAsApp).not.toHaveBeenCalled()
+    expect(dispatchedWorkflows).toHaveLength(2)
+    expect(dispatchedWorkflows.map(e => e.payload.inputs.package)).toEqual(['mingw-w64-gnutls', 'gnutls'])
```
Comment (rimrul) [inline __tests__/index.test.js:242]: Since we're expecting two dispatched workflows, shouldn't the comment body mention both?

**Review by rimrul — COMMENTED** [review]
One last remaining question, the rest looks good to me.

File: __tests__/index.test.js (hunk 3 above)
Comment (rimrul) [reply __tests__/index.test.js:154]: That looks good.

File: __tests__/index.test.js (hunk 4 above)
Comment (dscho) [reply __tests__/index.test.js:242]: I've added 5635fce742dea24d45eb2b32eca0593ad36f7876 to get _some_ coverage going. What do you think @rimrul?

Comment (rimrul) [reply __tests__/index.test.js:242]: Oh, right. That's two separate senzences in two different edits. I confused it with `/deploy` which has multiple workflows added to the comment at once.

**Pull request merged**